tools/integration-test/src/tests/client_upgrade.rs

```diff
@@ -0,0 +1,577 @@
+            )))
+        }
+    }
+}
+
+impl BinaryChainTest for ClientUpgradeTest {
+    fn run<
+        ChainA: ibc_test_framework::prelude::ChainHandle,
+        ChainB: ibc_test_framework::prelude::ChainHandle,
+    >(
+        &self,
+        _config: &ibc_test_framework::prelude::TestConfig,
+        _relayer: ibc_test_framework::prelude::RelayerDriver,
+        chains: ibc_test_framework::prelude::ConnectedChains<ChainA, ChainB>,
+    ) -> Result<(), ibc_test_framework::prelude::Error> {
+        let foreign_clients = chains.clone().foreign_clients;
+
+        let src_chain_config = chains.handle_b().config().unwrap();
+        let dst_chain_config = chains.handle_a().config().unwrap();
+
+        let upgraded_chain_id = chains.chain_id_a().0.clone();
+
+        let src_client_id = foreign_clients.client_id_b().0.clone();
+
+        // Create and send an chain upgrade proposal
+        let opts = UpgradePlanOptions {
+            dst_chain_config,
+            src_chain_config,
+            src_client_id,
+            amount: 10000000u64,
+            denom: "stake".to_string(),
+            height_offset: DELTA_HEIGHT,
+            upgraded_chain_id,
+            upgraded_unbonding_period: None,
+            upgrade_plan_name: "plan".to_string(),
+        };
+
+        build_and_send_ibc_upgrade_proposal(
+            chains.handle_a().clone(),
+            chains.handle_b().clone(),
+            &opts,
+        )
+        .map_err(Error::upgrade_chain)?;
+
+        // Wait for the proposal to be processed
+        std::thread::sleep(Duration::from_secs(2));
+
+        let driver = chains.node_a.chain_driver().0;
+
+        // Retrieve the height which should be used to upgrade the client
+        let upgrade_height = driver
+            .runtime
+            .block_on(query_upgrade_proposal_height(
+                &Uri::from_str(&driver.grpc_address()).unwrap(),
+                1,
+            ))
+            .unwrap();
+
+        let client_upgrade_height = Height::new(
+            foreign_clients.client_a_to_b.src_chain().id().version(),
+            upgrade_height,
+        )
+        .unwrap();
+
+        // Vote on the proposal so the chain will upgrade
+        vote_proposal(
+            driver.chain_id.as_str(),
+            &driver.command_path,
+            &driver.home_path,
+            &driver.rpc_listen_address(),
+        )?;
+
+        // The application height reports a height of 1 less than the height according to Tendermint
+        let target_reference_application_height = client_upgrade_height
+            .decrement()
+            .expect("Upgrade height cannot be 1");
+
+        let mut reference_application_latest_height = foreign_clients
+            .client_a_to_b
+            .src_chain()
+            .query_latest_height()
+            .ok()
+            .unwrap();
+
+        while reference_application_latest_height != target_reference_application_height {
```

Comment:
We may want to poll for a limited time so that the test don't run indefinitely when failed.

Reply:
This has been addressed in this commit https://github.com/informalsystems/hermes/pull/2795/commits/fa7ad9cc3e09eb4e8d1e8852d0976045f408ae78